### PR TITLE
Making ChunkMeta extensible

### PR DESCRIPTION
**Pull request by ykammeyer**
Adding $type entry, to indicate ChunkMeta class e.g.,
io.georocket.storage.ChunkMeta. IndexerVerticle and IndexedStoreCursor create ChunkMeta based on $type entry.

Transferred the elasticsearch document creation from IndexerVerticle to
ChunkMeta. The IndexerVerticle does not have to decide between ChunkMeta
classes, if different ChunkMeta should be supported by one IndexerVerticle
Implementation.

ChunkMeta can now add its ElasticsearchDocument Metadata to a given Map.

ChunkMeta can be directly created from a ElasticsearchDocument (Map or
JsonObject)
through a constructor.

### CODE REVIEW

File: georocket-server-api/src/main/java/io/georocket/storage/ChunkMeta.java [hunk 1]
```diff
@@ -33,6 +34,55 @@ public ChunkMeta(List<XMLStartElement> parents, int start, int end) {
   }
   
   /**
+   * Chunk Meta from JsonObject
+   * @param object The JsonObject containing the values for the ChunkMeta Object.
+   */
+  public ChunkMeta(JsonObject object){
+    JsonArray parentsArr = object.getJsonArray("parents");
+    parents = parentsArr.stream().map(e ->
```
Comment (michel-kraemer) [inline georocket-server-api/src/main/java/io/georocket/storage/ChunkMeta.java]: Should call other constructor `this(parents, start, end)`

File: georocket-server-api/src/main/java/io/georocket/storage/ChunkMeta.java [hunk 2]
```diff
@@ -33,6 +34,55 @@ public ChunkMeta(List<XMLStartElement> parents, int start, int end) {
...
+    end = object.getInteger("end");
+  }
+  
+  /**
+   * Chunk Meta from Map.
+   * Used to convert Elasticsearch hit to ChunkMeta
+   * @param source The Elasticsearch document
+   */
+  @SuppressWarnings("unchecked")
+  public ChunkMeta(Map<String, Object> source){
```
Comment (michel-kraemer) [inline georocket-server-api/src/main/java/io/georocket/storage/ChunkMeta.java]: The attributes you use here are only known to the concrete IndexerVerticle implementation. This code does not belong here.

File: georocket-server-api/src/main/java/io/georocket/storage/ChunkMeta.java [hunk 3]
```diff
@@ -103,7 +153,8 @@ public JsonObject toJsonObject() {
     return new JsonObject()
         .put("parents", ps)
         .put("start", start)
-        .put("end", end);
+        .put("end", end)
+        .put("$type", ChunkMeta.class.getName());
```
Comment (michel-kraemer) [inline georocket-server-api/src/main/java/io/georocket/storage/ChunkMeta.java]: `$type` should be added by the caller. The `ChunkMeta` should not need to care about this.

File: georocket-server-api/src/main/java/io/georocket/storage/ChunkMeta.java [hunk 4]
```diff
@@ -112,9 +163,18 @@ public JsonObject toJsonObject() {
...
-        XMLStartElement.fromJsonObject((JsonObject)e)).collect(Collectors.toList());
-    return new ChunkMeta(pl, obj.getInteger("start"), obj.getInteger("end"));
+    return new ChunkMeta(obj);
+  }
+  
+  /**
+   * Add own meta data to the given document
+   * @param doc The document to add the meta to
+   */
+  public void addMeta(Map<String, Object> doc) {
```
Comment (michel-kraemer) [inline georocket-server-api/src/main/java/io/georocket/storage/ChunkMeta.java]: Same here. These attributes belong to the indexer and not ChunkMeta.

File: georocket-server/src/main/java/io/georocket/index/IndexerVerticle.java [hunk 5]
```diff
@@ -587,36 +599,21 @@ private void addMeta(Map<String, Object> doc, ChunkMeta meta) {
...
-  /**
-   * Convert a list to an array. If the list is null the return value
-   * will also be null.
-   * @param list the list to convert
-   * @return the array or null if <code>list</code> is null
-   */
-  private String[] safeListToArray(List<String> list) {
-    if (list == null) {
-      return null;
+    ChunkMeta meta = null;
```
Comment (michel-kraemer) [inline georocket-server/src/main/java/io/georocket/index/IndexerVerticle.java]: Looks like duplicated code.

File: georocket-server/src/main/java/io/georocket/storage/indexed/IndexedStoreCursor.java [hunk 6]
```diff
@@ -115,7 +116,19 @@ private void handleResponse(JsonObject body) {
     for (int i = 0; i < count; ++i) {
       JsonObject hit = hits.getJsonObject(i);
       ids[i] = hit.getString("id");
-      metas[i] = ChunkMeta.fromJsonObject(hit);
+      
+      String metaType = hit.getString("$type");
```
Comment (michel-kraemer) [inline georocket-server/src/main/java/io/georocket/storage/indexed/IndexedStoreCursor.java]: Third time the same code.

**Comment by michel-kraemer** [comment]
Looks almost good to me. I'm just wondering why the ChunkMetaFactory accepts an error handler and doesn't just throw an exception. It is never used asynchronously, so all this additional code is not necessary IMHO.

**Comment by ykammeyer** [comment]
I though it would be an elegant way for the caller to log the exception without catching them. But you are right. I will change the method so it `throws Exception`. Thank you very much for your comments!

**Comment by michel-kraemer** [comment]
Thanks. Let me know when the PR is ready to be merged. And make sure the build runs. If you get an error on Travis CI you can either click the "Details" button or try to run the build on your local machine.

**Comment by michel-kraemer** [comment]
Some of the code (e.g. those parts dealing with adding the `$type` attribute, or the ChunkMetaFactory) are not necessary in the GeoRocket OSS. Please try to factor them out.

**Pull request merged**